src/Reconcile.js

```diff
@@ -0,0 +1,189 @@
+			return Component.update(host, parent, element, snapshot, Constant.object, Constant.update)
+		case Constant.thenable:
+			return Commit.thenable(host, parent, element, index, element.type = snapshot.type)
+		default:
+			children(host, element, element.children, snapshot.children)
+			props(element, element.props, snapshot.props)
+	}
+}
+
+/**
+ * @param {object} element
+ * @param {(string|number)} a
+ * @param {(string|number)} b
+ */
+export function content (element, a, b) {
+	if (a != b) {
+		Commit.content(element, element.children = b)
+	}
+}
+
+/**
+ * @param {object} element
+ * @param {object} a
+ * @param {object} b
+ */
+export function props (element, a, b) {
+	if (a !== b) {
+		Commit.props(element, object(a, element.props = b), element.xmlns, Constant.update)
+	}
+}
+
+/**
+ * @param {object} a
+ * @param {object} b
+ * @return {object?}
+ */
+export function object (a, b) {
+	var size = 0
+	var diff = {}
+
+	for (var key in a) {
+		if (!Utility.has(b, key)) {
+			diff[++size, key] = null
+		}
+	}
+
+	for (var key in b) {
+		var prev = a[key]
+		var next = b[key]
+
+		if (prev !== next) {
+			if (typeof next !== 'object' || next === null) {
+				diff[++size, key] = next
+			} else if (next = object(prev || {}, next)) {
+				diff[++size, key] = next
+			}
+		}
+	}
+
+	if (size) {
+		return diff
+	}
+}
+
+/**
+ * @param {object} host
+ * @param {object} parent
+ * @param {object} a
+ * @param {number} b
+ */
+export function children (host, parent, a, b) {
+	var prev = parent
+	var next = parent
+	var alen = a.length
+	var blen = b.length
+	var aend = alen - 1
+	var bend = blen - 1
+	var min = bend > aend ? aend : bend
+	var max = bend > aend ? bend : aend
+	var idx = 0
+	var ptr = 0
+
+	// step 1, prefix/suffix
+	outer: {
+		while ((prev = a[idx]).key === (next = b[idx]).key) {
+			update(host, parent, prev, next, idx)
+			if (++idx > min) { break outer }
+		}
+		while ((prev = a[aend]).key === (next = b[bend]).key) {
+			update(host, parent, a[max] = prev, next, bend)
+			if (--aend, --bend, --max, idx > --min) { break outer }
+		}
+	}
+
+	// step 2, mount/unmount/sort
+	if (idx > aend++) {
+		if (idx <= bend++) {
+			if (bend < blen) do {
+				Commit.mount(parent, a[--bend] = Commit.create(host, parent, b[bend], Constant.create), a[bend + 1])
+			} while (idx < bend) else do {
+				Commit.mount(parent, a[--bend] = Commit.create(host, parent, b[bend], Constant.create), null)
+			} while (idx < bend)
+		}
+	} else if ((ptr = idx) > bend++) {
+		do { Commit.unmount(parent, Commit.destroy(parent, a[idx++])) } while (idx < aend)
+		a.splice(ptr, aend - bend)
+	} else {
+		sort(host, parent, a, b, idx, aend, bend)
+	}
+}
+
+/**
+ * @param {object} host
+ * @param {object} parent
+ * @param {object} a
+ * @param {object} b
+ * @param {number} idx
+ * @param {number} aend
+ * @param {number} bend
+ */
+export function sort (host, parent, a, b, idx, aend, bend) {
+	var aidx = idx
+	var bidx = idx
+	var akey = []
```

Comment:
These should probably be objects, arrays have the benefit when all keys are numbers(small) but this leaves a lot to the imagination whether an engine is eager as opposed to lazy when you do `akey[1000] = 1` and create space for 1000 items when you only have 4 items and one of them has a key of `1000`. In any case even if they are numeric keys number keys in objects are stored differently from other keys so you might arrive at what you meant to archive in the first place without the trade offs.